### PR TITLE
Modernize code to c++11

### LINKED ISSUE #17600: Modernize more code to use c++11
Modernize more code to use c++11. Specifically functions now found in `<string>`. `atoi` should move to `std::stoi`, `atof` should move to `std::stod`

### PR DESCRIPTION
**Pull request by lottiegreenwood**
Modernize more code to use c++11. Specifically functions now found in `<string>`. `atoi` should move to `std::stoi`, `atof` should move to `std::stod`.

**To test:**

Check tests all pass, functions as normal.

Fixes #17600.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/src/FindDetectorsPar.cpp [hunk 1]
```diff
@@ -625,7 +625,7 @@ FindDetectorsPar::get_ASCII_header(std::string const &fileName,
     file_descriptor.data_start_position =
         data_stream.tellg(); // if it is PHX or PAR file then the data begin
                              // after the first line;
-    file_descriptor.nData_records = atoi(&BUF[0]);
+    file_descriptor.nData_records = std::stoi(&BUF[0]);
```
Comment (quantumsteve) [inline Framework/DataHandling/src/FindDetectorsPar.cpp]: While you're modernizing, change `&BUF[0]` to [`BUF.data()`](http://en.cppreference.com/w/cpp/container/vector/data).

File: Framework/DataHandling/src/LoadFullprofResolution.cpp [hunk 2]
```diff
@@ -540,12 +520,9 @@ void LoadFullprofResolution::parseResolutionStrings(
...
-        tempdb = atof(terms[1].c_str());
-        parammap["Sig2"] = sqrt(tempdb);
-        tempdb = atof(terms[2].c_str());
-        parammap["Sig1"] = sqrt(tempdb);
-        tempdb = atof(terms[3].c_str());
-        parammap["Sig0"] = sqrt(tempdb);
+        parammap["Sig2"] = parseDoubleValue(terms[1], "Sig2");
+        parammap["Sig1"] = parseDoubleValue(terms[2], "Sig1");
+        parammap["Sig0"] = parseDoubleValue(terms[3], "Sig0");
       }
```
Comment (quantumsteve) [inline Framework/DataHandling/src/LoadFullprofResolution.cpp]: Did you drop a `sqrt` here?

File: Framework/API/src/IFunction.cpp [hunk 3]
```diff
@@ -901,7 +901,7 @@ void IFunction::setMatrixWorkspace(
               if (fitParam.getConstraintPenaltyFactor().compare("")) {
                 try {
                   double penalty =
-                      atof(fitParam.getConstraintPenaltyFactor().c_str());
+                      std::stof(fitParam.getConstraintPenaltyFactor().c_str());
```
Comment (quantumsteve) [inline Framework/API/src/IFunction.cpp]: 1. `std::stof` returns a float, but  `penalty` is a double.
2. remove `.c_str()`

File: Framework/API/src/MultiPeriodGroupWorker.cpp [hunk 4]
```diff
@@ -280,15 +280,15 @@ void MultiPeriodGroupWorker::validateMultiPeriodGroupInputs(
                 currentGroup->getItem(j));
         Property *nPeriodsProperty =
             currentNestedWS->run().getLogData("nperiods");
-        size_t nPeriods = atoi(nPeriodsProperty->value().c_str());
+        size_t nPeriods = std::stoi(nPeriodsProperty->value());
         if (nPeriods != benchMarkGroupSize) {
```
Comment (quantumsteve) [inline Framework/API/src/MultiPeriodGroupWorker.cpp]: Would [std::stoul](http://en.cppreference.com/w/cpp/string/basic_string/stoul) be more precise?

File: Framework/API/src/MultiPeriodGroupWorker.cpp [hunk 5]
```diff
@@ -280,15 +280,15 @@ void MultiPeriodGroupWorker::validateMultiPeriodGroupInputs(
...
           throw std::runtime_error("Missmatch between nperiods log and the "
                                    "number of workspaces in the input group: " +
                                    vecMultiPeriodGroups[i]->getName());
         }
         Property *currentPeriodProperty =
             currentNestedWS->run().getLogData("current_period");
-        size_t currentPeriod = atoi(currentPeriodProperty->value().c_str());
+        size_t currentPeriod = std::stoi(currentPeriodProperty->value());
```
Comment (quantumsteve) [inline Framework/API/src/MultiPeriodGroupWorker.cpp]: `std::stoul`?

File: Framework/Algorithms/src/GenerateEventsFilter.cpp [hunk 6]
```diff
@@ -312,7 +312,7 @@ void GenerateEventsFilter::processInputTime() {
     m_startTime = DateAndTime(s_inpt0);
   } else {
     // Relative time in double.
-    double inpt0 = atof(s_inpt0.c_str());
+    double inpt0 = std::stof(s_inpt0.c_str());
```
Comment (quantumsteve) [inline Framework/Algorithms/src/GenerateEventsFilter.cpp]: std::stof returns a float, but inpt0 is a double.

File: Framework/Algorithms/src/GenerateEventsFilter.cpp [hunk 7]
```diff
@@ -332,7 +332,7 @@ void GenerateEventsFilter::processInputTime() {
     m_stopTime = DateAndTime(s_inptf);
   } else {
     // Relative time in double
-    double inptf = atof(s_inptf.c_str());
+    double inptf = std::stof(s_inptf.c_str());
```
Comment (quantumsteve) [inline Framework/Algorithms/src/GenerateEventsFilter.cpp]: std::stof returns a float, but inptf is a double.

File: Framework/Algorithms/src/PlotAsymmetryByLogValue.cpp [hunk 8]
```diff
@@ -228,8 +228,8 @@ void PlotAsymmetryByLogValue::checkProperties(size_t &is, size_t &ie) {
   // Parse run names and get the number of runs
   parseRunNames(firstFN, lastFN, m_filenameBase, m_filenameExt,
                 m_filenameZeros);
-  is = atoi(firstFN.c_str()); // starting run number
-  ie = atoi(lastFN.c_str());  // last run number
+  is = std::stoi(firstFN); // starting run number
+  ie = std::stoi(lastFN);  // last run number
```
Comment (quantumsteve) [inline Framework/Algorithms/src/PlotAsymmetryByLogValue.cpp]: `std::stoul`?

File: Framework/Crystal/src/LoadIsawPeaks.cpp [hunk 9]
```diff
@@ -240,26 +240,26 @@ DataObjects::Peak LoadIsawPeaks::readPeak(PeaksWorkspace_sptr outWS,
...
+  wl = std::stod(getWord(in, false), nullptr);
+  UNUSED_ARG(std::stod(getWord(in, false), nullptr)); // D
+  IPK = std::stod(getWord(in, false), nullptr);
 
-  Inti = strtod(getWord(in, false).c_str(), nullptr);
-  SigI = strtod(getWord(in, false).c_str(), nullptr);
-  static_cast<void>(atoi(getWord(in, false).c_str())); // iReflag
+  Inti = std::stod(getWord(in, false), nullptr);
+  SigI = std::stod(getWord(in, false), nullptr);
+  static_cast<void>(std::stoi(getWord(in, false))); // iReflag
```
Comment (quantumsteve) [inline Framework/Crystal/src/LoadIsawPeaks.cpp]: change to `UNUSED_ARG`?

File: Framework/DataHandling/src/LoadFullprofResolution.cpp [hunk 10]
```diff
@@ -654,13 +620,31 @@ void LoadFullprofResolution::parseBankLine(string line, double &cwl,
...
       if (!candidate.empty()) {
-        cwl = atof(candidate.c_str());
+        cwl = parseDoubleValue(candidate);
         break;
       }
     }
   }
 }
 
+double LoadFullprofResolution::parseDoubleValue(const std::string &value,
```
Comment (quantumsteve) [inline Framework/DataHandling/src/LoadFullprofResolution.cpp]: Does this need to be a member function? Would it be better as a static function inside the implementation file?

File: Framework/DataHandling/src/SaveGSASInstrumentFile.cpp [hunk 11]
```diff
@@ -235,7 +235,7 @@ ChopperConfiguration::parseStringDbl(const string &instring) const {
   vector<double> vecdouble;
   for (auto &str : strs) {
     if (!str.empty()) {
-      double item = atof(str.c_str());
+      double item = stod(str.c_str());
```
Comment (quantumsteve) [inline Framework/DataHandling/src/SaveGSASInstrumentFile.cpp]: `std::stod`?

File: Framework/DataObjects/src/AffineMatrixParameterParser.cpp [hunk 12]
```diff
@@ -41,7 +41,7 @@ AffineMatrixParameter *AffineMatrixParameterParser::createParameter(
       nCols = vecStrCols.size();
       col_it = vecStrCols.begin();
       while (col_it != vecStrCols.end()) {
-        coord_t val = coord_t(atof(col_it->c_str()));
+        coord_t val = coord_t(std::stof(col_it->c_str()));
```
Comment (quantumsteve) [inline Framework/DataObjects/src/AffineMatrixParameterParser.cpp]: Remove the c-style cast

File: Framework/MDAlgorithms/src/ConvertSpiceDataToRealSpace.cpp [hunk 13]
```diff
@@ -594,7 +593,7 @@ IMDEventWorkspace_sptr ConvertSpiceDataToRealSpace::createDataMDWorkspace(
 
   for (const auto &thisWorkspace : vec_ws2d) {
     short unsigned int runnumber = static_cast<short unsigned int>(
-        atoi(thisWorkspace->run().getProperty("run_number")->value().c_str()));
+        std::stoi(thisWorkspace->run().getProperty("run_number")->value()));
```
Comment (quantumsteve) [inline Framework/MDAlgorithms/src/ConvertSpiceDataToRealSpace.cpp]: should runnumber really be casted to an `unsigned short`???

File: Framework/MDAlgorithms/src/IntegratePeaksCWSD.cpp [hunk 14]
```diff
@@ -510,11 +510,11 @@ std::map<int, signal_t> IntegratePeaksCWSD::getMonitorCounts() {
     std::string run_str = expinfo->run().getProperty("run_number")->value();
     g_log.information() << "run number of exp " << iexpinfo << " is " << run_str
                         << "\n";
-    int run_number = atoi(run_str.c_str());
+    int run_number = std::stoi(run_str);
     // FIXME - HACK FOE HB3A
     run_number = run_number % 1000;
     std::string mon_str = expinfo->run().getProperty("monitor")->value();
-    signal_t monitor = static_cast<signal_t>(atoi(mon_str.c_str()));
+    signal_t monitor = static_cast<signal_t>(std::stoi(mon_str));
```
Comment (quantumsteve) [inline Framework/MDAlgorithms/src/IntegratePeaksCWSD.cpp]: I believe `signal_t` == `double`. Should one use directly use `std::stod`?

File: Framework/MDAlgorithms/src/IntegratePeaksCWSD.cpp [hunk 15]
```diff
@@ -537,12 +537,12 @@ std::map<int, double> IntegratePeaksCWSD::getMeasureTime() {
...
         m_inputWS->getExperimentInfo(static_cast<uint16_t>(iexpinfo));
     std::string run_str = expinfo->run().getProperty("run_number")->value();
-    int run_number = atoi(run_str.c_str());
+    int run_number = std::stoi(run_str);
 
     // FIXME - HACK FOE HB3A
     run_number = run_number % 1000;
     std::string duration_str = expinfo->run().getProperty("duration")->value();
-    double duration = static_cast<double>(atof(duration_str.c_str()));
+    double duration = static_cast<double>(std::stod(duration_str));
```
Comment (quantumsteve) [inline Framework/MDAlgorithms/src/IntegratePeaksCWSD.cpp]: remove `static_cast`

File: MantidPlot/src/ConfigDialog.cpp [hunk 16]
```diff
@@ -700,7 +700,7 @@ void ConfigDialog::initMantidPage() {
   const std::string ignoreParaViewProperty = "paraview.ignore";
   bool ignoreParaView =
       cfgSvc.hasProperty(ignoreParaViewProperty) &&
-      bool(atoi(cfgSvc.getString(ignoreParaViewProperty).c_str()));
+      bool(std::stoi(cfgSvc.getString(ignoreParaViewProperty)));
```
Comment (quantumsteve) [inline MantidPlot/src/ConfigDialog.cpp]: Remove the c-style cast by comparing the result with `0`

**Review by quantumsteve — DISMISSED** [review]
This change will significantly improve validation of input from IDFs by ensuring failed conversions are made visible to the user.

There are several implicit conversions that still need to be cleaned up before merging.

File: Framework/DataHandling/src/LoadFullprofResolution.cpp (hunk 10 above)
Comment (peterfpeterson) [inline Framework/DataHandling/src/LoadFullprofResolution.cpp]: It is a member so the warning gives a meaningful message associated with the algorithm

File: Framework/MDAlgorithms/src/ConvertSpiceDataToRealSpace.cpp (hunk 13 above)
Comment (quantumsteve) [inline Framework/MDAlgorithms/src/ConvertSpiceDataToRealSpace.cpp]: Wow, didn't realize we only use 16-bit integers for the run number.

File: MantidPlot/src/ConfigDialog.cpp [hunk 17]
```diff
@@ -700,7 +700,7 @@ void ConfigDialog::initMantidPage() {
   const std::string ignoreParaViewProperty = "paraview.ignore";
   bool ignoreParaView =
       cfgSvc.hasProperty(ignoreParaViewProperty) &&
-      bool(std::stoi(cfgSvc.getString(ignoreParaViewProperty)));
+      (std::stoi(cfgSvc.getString(ignoreParaViewProperty)) == 0);
```
Comment (quantumsteve) [inline MantidPlot/src/ConfigDialog.cpp]: Is this reversed?

**Comment by quantumsteve** [comment]
:shipit:

**Pull request merged**